File: python/p026.py
#!/usr/bin/python


def prime_sieve(limit=125000):
    if limit % 2 != 0:
        limit += 1
    sieve = [True] * limit
    sieve[0], sieve[1] = [False] * 2
    for ind, val in enumerate(sieve):
        if val is True:
            sieve[ind * 2::ind] = [False] * (((limit - 1) // ind) - 1)
    return sieve


def primes(limit=125000):
    return [i for i, v in enumerate(prime_sieve(limit)) if v is True]


limit = 1000

for d in primes(limit)[::-1]:
    period = 1
    while pow(10, period, d) != 1:
        period += 1
    if d - 1 == period:
        break

print(d)

File: rust/src/bin/p026.rs

extern crate num;
use num::bigint::ToBigUint;


fn prime_sieve(mut limit: usize) -> Vec<bool> {
    if limit % 2 != 0 {
        limit += 1;
    }
    let mut sieve = vec![true; limit];
    sieve[0] = false; sieve[1] = false;
    let mut i = 0;
    while i < sieve.len() {
        if sieve[i] == true {
            let mut j = i * i;
            while j < sieve.len() {
                sieve[j] = false;
                j += i;
            }
        }
        i += 1;
    }
    sieve
}


fn primes(limit: usize) -> Vec<usize> {
    prime_sieve(limit).iter().enumerate().filter(|&(_, v)| *v == true).map(|(i, _)| i).collect()
}


fn main() {

    let mut res = 0;
    let limit = 1000;
    let mut period: usize;

    for d in primes(limit).iter().rev() {
        res = *d;
        period = 1;
        while num::pow(10.to_biguint().unwrap(), period) % d.to_biguint().unwrap() != 1.to_biguint().unwrap() {
            period += 1;
        }
        if d - 1 == period {
            break;
        }
    }

    println!("{}", res);

}
